src/render.rs

```diff
@@ -1,145 +1,119 @@
+                self.write_channel_message(0xB0, channel.into(), &[control.into(), value.into()])?;
+            }
+            MidiMessage::ProgramChange(channel, program) => {
+                self.write_channel_message(0xC0, channel.into(), &[program.into()])?;
+            }
+            MidiMessage::ChannelPressure(channel, value) => {
+                self.write_channel_message(0xD0, channel.into(), &[value.into()])?;
+            }
+            MidiMessage::PitchBendChange(channel, value) => {
+                let (value_lsb, value_msb) = value.into();
+                self.write_channel_message(0xE0, channel.into(), &[value_lsb, value_msb])?;
+            }
+            MidiMessage::QuarterFrame(value) => {
+                self.transport.write(0xF1)?;
+                self.transport.write(value.into())?;
+                self.running_state = None;
+            }
+            MidiMessage::SongPositionPointer(value) => {
+                let (value_lsb, value_msb) = value.into();
+                self.transport.write(0xF2)?;
+                self.transport.write(value_lsb)?;
+                self.transport.write(value_msb)?;
+                self.running_state = None;
+            }
+            MidiMessage::SongSelect(value) => {
+                self.transport.write(0xF3)?;
+                self.transport.write(value.into())?;
+                self.running_state = None;
+            }
+            MidiMessage::TuneRequest => {
+                self.transport.write(0xF6)?;
+                self.running_state = None;
+            }
+            MidiMessage::TimingClock => {
+                self.transport.write(0xF8)?;
+            }
+            MidiMessage::Start => {
+                self.transport.write(0xFA)?;
+            }
+            MidiMessage::Continue => {
+                self.transport.write(0xFB)?;
+            }
+            MidiMessage::Stop => {
+                self.transport.write(0xFC)?;
+            }
+            MidiMessage::ActiveSensing => {
+                self.transport.write(0xFE)?;
+            }
+            MidiMessage::Reset => {
+                self.transport.write(0xFF)?;
+            }
+        }
+        Ok(())
     }
 
-    #[test]
-    fn render_ok() {
-        let mut buf3 = [0, 0, 0];
-        let mut buf100 = [0; 100];
-        for v in (*TEST_1BYTE).iter() {
-            assert_eq!(1, v.render_slice(&mut buf3), "{:?}", v);
-            assert_eq!(1, v.render_slice(&mut buf100), "{:?}", v);
+    fn write_channel_message(
+        &mut self,
+        status_msb: u8,
+        channel: u8,
+        data: &[u8],
+    ) -> Result<(), T::Error> {
+        let status = status_msb + channel;
+        // If the last command written had the same status/channel, the MIDI protocol allows us to
+        // omit sending the status byte again.
+        if self.running_state != Some(status) {
+            self.transport.write(status)?;
         }
```

Comment:
I *think* you might need to move this logic to the `transport` layer?

I need to refresh myself on the USB implementation but I don't believe it does the byte omission as it has no real need for it and has to write packet headers based on the content. For instance, NoteOn messages always have a specific header and 3 bytes of data. As is, the USB implementation we have simply uses the `MidiRenderSlice` but we could potentially make that a `Transport` implementation to consolidate the logic?

https://usb.org/sites/default/files/midi10.pdf

maybe `Transport` could either have a boolean with write to indicate an optional omit or, you could add an additional method or something that defaults to simply drop the byte

```rust
pub trait MidiTransport {
    type Error;
    fn write(&mut self, value: u8) -> Result<(), Self::Error>;
    fn write_optional_status(&mut self, _value: u8) -> Result<(), Self::Error> {
      //default, drop status
      Ok(())
    }
}
```

Reply:
Ah, thanks for correcting me on that. TIL that running status is not allowed in USB Midi :-). I think a flag would be the best solution but we can put that in the Renderer itself. I think if I make it a const generic it can be optimized out by the compiler.

Reply:
I was just wondering, maybe you could make the transport slice oriented and still use the const generic in the renderer so you can fully merge these two renderers and not lose the optimization for slice rendering? I assume there is no downside for byte oriented transports to process `n` bytes at a time instead of `1`

Reply:
I don't think we can solve that problem by changing the transport in this way. Rendering with running state requires the renderer to keep state between messages, the slice renderer does not do that by design. You can't solve that by pushing rendering logic to the transport.

Reply:
We can probably make a render function that takes the previous running state as a parameter and returns the next running state and use that in both implementations, for the slice renderer you can just pass in None and ignore the result. That would make both of the renderers slightly less efficient but not much.

But to be honest I don't know if we still need the render_slice implementation. Especially if we want to implement writing sysex which would require writing big blobs of data, that wont fit in the 3 bytes buffer of the slice renderer.

Reply:
this is roughly what I was thinking, leaving the transport unaware of the running status:

```rust
pub trait MidiTransport {
    type Error;

    /// Write a slice to the midi transport layer
    fn write(&mut self, value: &[u8]) -> Result<(), Self::Error>;
}

/// The MidiRenderer takes MIDI messages and writes them to the underlying transport
#[derive(Debug)]
pub struct MidiRenderer<T, const OMIT_REPEATED: bool = true> {
    transport: T,
    running_state: Option<u8>,
}

impl<T, const OMIT_REPEATED: bool> MidiRenderer<T, OMIT_REPEATED> {
    fn write_channel_message(
        &mut self,
        status_msb: u8,
        channel: u8,
        data: &[u8],
    ) -> Result<(), T::Error> {
        let status = status_msb + channel;

        // If the last command written had the same status/channel, the MIDI protocol allows us to
        // omit sending the status byte again.
        if OMIT_REPEATED && self.running_state == Some(status) {
            self.transport.write(data)?;
        } else if data.len() == 2 {
            let out = [status, data[0], data[1]];
            self.transport.write(out)?;
        } else {
            let out = [status, data[0]];
            self.transport.write(out)?;
        }

        // Store running state so it can be re-used later
        self.running_state = Some(status);

        Ok(())
    }

    pub fn render(&mut self, message: &MidiMessage) -> Result<(), T::Error> {
        match *message {
            MidiMessage::NoteOn(channel, note, velocity) => {
                self.write_channel_message(0x90, channel.into(), &[note.into(), velocity.into()])?;
            }
            ...
            MidiMessage::QuarterFrame(value) => {
                self.transport.write(&[0xF1, value.into()])?;
                self.running_state = None;
            }
            ...
        }
    }
}

```

Reply:
Ah yeah, that is pretty much exactly what I just committed :-D I'll add the default value for the const generic too.

Reply:
If you notice, my code example above uses slices, not individual bytes, so it could fully replace the slice renderer without requiring bounds checking for each byte written to the transport

Reply:
Ah, sorry, it was late yesterday so I missed that. I'm not opposed to doing it this way but I'm not really clear what the benefit of using buffers would be. Does the usb implementation need that?
Maybe I should go through that code again next weekend. It's been a couple of months since I looked at that implementation.

Reply:
I forget which branch is my most recent in the USB midi detail but here is a rough idea:
https://github.com/x37v/usbd-midi/blob/xnor/midi-types-reorg-rebase/src/event_packet.rs#L18-L29

To send a USB packet, we render to a slice of a `[u8; 4]`

Reply:
Ah yeah, that totally makes sense.

Reply:
I think this should work better with Midi over USB.
I also moved a bunch of rendering tests from the embedded-midi crate and adapted them.